### PR TITLE
Github pipelines and utils for running CI on parallel works

### LINKED ISSUE #3006: Setup and run CI with self-hosted GitHub Runners on Parallel Works
### What new functionality do you need?

As a code manager I want the ability to seamlessly run a CI  test suite on virtual cluster in Parallel Works.

### What are the requirements for the new functionality?

Using self-hosted GitHub Runners configured on and AWS virtual cluster run and end-to-end CI test suite from an action directly on GitHub.

### Acceptance Criteria

 - [x] Setup a self hosted runner and have it self provision on an AWS virtual cluster
 - [x] Writ a GitHub Actions Pipeline to run a full CI test
 - [x] Test the Pipeline on AWS from a dispatch action on GitHub
 - [ ] Perform a self test after the ./github/workflow udates are checkin to the authoritative global-workflow repo

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by TerrenceMcGuinness-NOAA**
# Description

This PR has the GitHub Pipeline script in the `github/workflows` directory for running CI tests
to be preformed an AWS virtual cluster.  It is setup to be launched from the dispatch action from the Actions tab.

For now it will only run C48_ATM 

Resolves #3006 

Once the yaml pipeline is in `.github/workflows` directory of the default branch we can test it against [PR 2977](https://github.com/NOAA-EMC/global-workflow/pull/2977) which may be needed to build on Parallel Works Centos AWS.

Code managers can check to see if the self-hosted runner [globalworkflow_parallelworks](https://github.com/NOAA-EMC/global-workflow/settings/actions/runners/22) is up and ready by checking the [Running](https://github.com/NOAA-EMC/global-workflow/settings/actions/runners) Settings.

In pending work we should also be able spin up the cluster on demand from GitHub as well.

# Type of change
- [ ] Bug fix (fixes something broken)
- [ ] New feature (adds functionality)
- [x] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics

- Is this a breaking change (a change in existing functionality)? YES/NO
- Does this change require a documentation update? YES/NO
- Does this change require an update to any of the following submodules? YES/NO (If YES, please add a link to any PRs that are pending.)
  - [ ] EMC verif-global 
  - [ ] GDAS 
  - [ ] GFS-utils 
  - [ ] GSI 
  - [ ] GSI-monitor 
  - [ ] GSI-utils 
  - [ ] UFS-utils 
  - [ ] UFS-weather-model 
  - [ ] wxflow 

# How has this been tested?

# Checklist
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my code, including function, input, and output descriptions
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [x] This change is covered by an existing CI test or a new one has been added
- [ ] I have made corresponding changes to the system documentation if necessary

### CODE REVIEW

File: ci/scripts/utils/parallel_works/provision_runner.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+   echo "actions-runner is already running"
+   exit
+fi
+
+cp /contrib/${CI_USER}/SETUP/actions-runner_${OS_NAME}.tar.gz ${HOME}
+cd $HOME || exit
+tar -xf "actions-runner_${OS_NAME}.tar.gz"
+cd actions-runner || exit
+d=$(date +%Y-%m-%d-%H:%M)
+nohup ./run.sh >& "run_noup${d}.log" &
```
Comment (aerorahul) [inline ci/scripts/utils/parallel_works/provision_runner.sh]:
```suggestion
nohup ./run.sh >& "run_nohup${d}.log" &
```

File: ci/scripts/utils/parallel_works/provision_runner.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  exit 1
+fi
+
+running=$(pgrep -u ${USER} run-helper -c) || true
+if [[ "${running}" -gt 0 ]]; then
+   echo "actions-runner is already running"
+   exit
+fi
+
+cp /contrib/${CI_USER}/SETUP/actions-runner_${OS_NAME}.tar.gz ${HOME}
```
Comment (aerorahul) [inline ci/scripts/utils/parallel_works/provision_runner.sh]: What is in this `tar.gz`?

File: .github/workflows/globalworkflow-ci-pw-centos.yaml [hunk 3]
```diff
@@ -15,28 +15,25 @@ on: [workflow_dispatch]
 #         └── ${pslot}
 env:
   TEST_DIR: ${{ github.workspace }}/${{ github.run_id }}
-  MACHINE_ID: orion
+  MACHINE_ID: noaacloud
 
 jobs:
   checkout-build-link:
```
Comment (aerorahul) [inline .github/workflows/globalworkflow-ci-pw-centos.yaml]: Can we have 3 steps? `checkout`, `build` and `link` OR 2 `checkout` and `build_link`.

File: .github/workflows/globalworkflow-ci-pw-centos.yaml [hunk 4]
```diff
@@ -45,42 +42,41 @@ jobs:
 
   create-experiments:
     needs: checkout-build-link
-    runs-on: [self-hosted, orion-ready]
+    runs-on: [self-hosted, aws, parallelworks, centos]
```
Comment (aerorahul) [inline .github/workflows/globalworkflow-ci-pw-centos.yaml]: does `runs-on` need to be on every step even if it is described on line 22?

File: ci/scripts/utils/parallel_works/provision_runner.sh [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  exit 1
+fi
+
+running=$(pgrep -u "${USER}" run-helper -c) || true
+if [[ "${running}" -gt 0 ]]; then
+   echo "actions-runner is already running"
+   exit
+fi
+
+cp "/contrib/${CI_USER}/SETUP/actions-runner_${OS_NAME}.tar.gz ${HOME}"
```
Comment (aerorahul) [inline ci/scripts/utils/parallel_works/provision_runner.sh]:
```suggestion
cp "/contrib/${CI_USER}/SETUP/actions-runner_${OS_NAME}.tar.gz" "${HOME}/"
```

**Review by aerorahul — COMMENTED** [review]
looks good, just some minor comments.

**Review by aerorahul — DISMISSED** [review]
looks good.
still would like to know if the `runs-on` is needed in every step even if it is the same fo all the jobs.

File: .github/workflows/globalworkflow-ci-pw-centos.yaml
Comment (aerorahul) [inline .github/workflows/globalworkflow-ci-pw-centos.yaml:1]: would it be possible to rename this file to `pw_aws_centos.yaml`

**Comment by TerrenceMcGuinness-NOAA** [comment]
@aerorahul Yes, the runs-on attribute must be specified for each job in a GitHub Actions workflow. This attribute defines the type of runner that the job will run on. Each job can have different requirements for the runner, so it is necessary to specify runs-on for each job individually. Again, note it's defined on every job level, not step.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Review by aerorahul — APPROVED** [review]
lgtm. thanks for including the changes.

**Pull request merged**